executor/aggfuncs/func_value_test.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs_test
+
+import (
+	. "github.com/pingcap/check"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/executor/aggfuncs"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+func getEvaluatedMemDelta(row *chunk.Row, dataType *types.FieldType) (memDelta int64) {
+	memDelta = 0
+	switch dataType.Tp {
+	case mysql.TypeString:
+		memDelta = int64(len(row.GetString(0)))
+	case mysql.TypeJSON:
+		memDelta = int64(len(row.GetJSON(0).Value))
+	}
+	return
+}
+
+func firstValueEvaluateRowUpdateMemDeltaGens(srcChk *chunk.Chunk, dataType *types.FieldType) (memDeltas []int64, err error) {
+	memDeltas = make([]int64, 0)
+	if srcChk.NumRows() > 0 {
+		row := srcChk.GetRow(0)
+		memDeltas = append(memDeltas, getEvaluatedMemDelta(&row, dataType))
+	}
+	for i := 1; i < srcChk.NumRows(); i++ {
+		memDeltas = append(memDeltas, int64(0))
+	}
+	return memDeltas, nil
+}
+
+func lastValueEvaluateRowUpdateMemDeltaGens(srcChk *chunk.Chunk, dataType *types.FieldType) (memDeltas []int64, err error) {
+	memDeltas = make([]int64, 0)
+	for i := 0; i < srcChk.NumRows(); i++ {
+		row := srcChk.GetRow(0)
+		memDeltas = append(memDeltas, getEvaluatedMemDelta(&row, dataType))
+	}
+	return memDeltas, nil
+}
+
+func nthValueEvaluateRowUpdateMemDeltaGens(nth int) updateMemDeltaGens {
```

Comment:
Sure. ```nthValueEvaluateRowUpdateMemDeltaGens``` is actually a high-order function which can generate  ```firstValueEvaluateRowUpdateMemDeltaGens```, but I don't  think it can make another  ```lastValueEvaluateRowUpdateMemDeltaGens```. So I only removed ```firstValueEvaluateRowUpdateMemDeltaGens```. PTAL.

Reply:
Why it can not make `lastValueEvaluateRowUpdateMemDeltaGens`?

Reply:
The logic in the 3 ```MemDeltaGens``` functions are corresponding to the 3 different ```UpdatePartialResult``` functions for ```firstValue```, ```lastValue``` and ```nthValue```
From the code in executor/aggfuncs/func_value.go, the 3 ```UpdatePartialResult``` functions are compared as follow:
```
//first: if rowsInGroup is not empty for the first time, memDelta is not zero(for string and json).
p := (*partialResult4FirstValue)(pr)
if p.gotFirstValue {
	return 0, nil
}
if len(rowsInGroup) > 0 {
	p.gotFirstValue = true
	memDelta, err = p.evaluator.evaluateRow(sctx, v.args[0], rowsInGroup[0])
	if err != nil {
		return 0, 
}
return memDelta, nil
```
```
//last: if rowsInGroup is not empty, memDelta is always not zero(for string and json).
p := (*partialResult4LastValue)(pr)
if len(rowsInGroup) > 0 {
	p.gotLastValue = true
	memDelta, err = p.evaluator.evaluateRow(sctx, v.args[0], rowsInGroup[len(rowsInGroup)-1])
	if err != nil {
		return 0, err
	}
}
return memDelta, nil
```
```
//nth: if p.seenRows(which is accumulated rowsInGroups) reached v.nth for the first time, 
//memDelta is not zero(for string and json). Otherwise it is zero.
if v.nth == 0 {
	return 0, nil
}
p := (*partialResult4NthValue)(pr)
numRows := uint64(len(rowsInGroup))
if v.nth > p.seenRows && v.nth-p.seenRows <= numRows {
	memDelta, err = p.evaluator.evaluateRow(sctx, v.args[0], rowsInGroup[v.nth-p.seenRows-1])
	if err != nil {
		return 0, err
	}
}
p.seenRows += numRows
return memDelta, nil
```
So for unit tests consisting 3 rows of type string, their corresponding memDelta should be like:
first:               1, 0, 0
second(nth): 0, 1, 0
last:                1, 1, 1
Above is what I understood from the original code. If I was wrong, please let me know. Thanks!

Reply:
The difference between ```firstValueEvaluateRowUpdateMemDeltaGens``` and ```lastValueEvaluateRowUpdateMemDeltaGens``` was verified by @XuHuaiyu. So I think it's better to keep ```nth``` and ```last``` as two separate functions.

Reply:
you are right, `lastValueEvaluateRowUpdateMemDeltaGens` needs to be assigned multiple times, while `firstValueEvaluateRowUpdateMemDeltaGens` and `nthValueEvaluateRowUpdateMemDeltaGens` only need to be assigned once, so it is ok to write separately.